src/Typedefs.idr

```diff
@@ -3,34 +3,129 @@ module Typedefs
 import Data.Fin
 import Data.Vect
 
-import Types
+import Names
 
 %default total
 %access public export
 
-||| Type definition
-||| @n The number of type variables in the type
-data TDef : (n:Nat) -> Type where
-  ||| The empty type
-  T0    :                                      TDef n
+mutual
+  ||| Type definition
+  ||| @n The number of type variables in the type
+  data TDef : (n:Nat) -> Type where
+    ||| The empty type
+    T0    :                                          TDef n
+
+    ||| The unit type
+    T1    :                                          TDef n
+
+    ||| The coproduct type
+    TSum  :             Vect (2 + k) (TDef n)     -> TDef n
+
+    ||| The product type
+    TProd :             Vect (2 + k) (TDef n)     -> TDef n
+
+    ||| A type variable
+    ||| @i De Bruijn index
+    TVar  :             (i:Fin (S n))             -> TDef (S n)
+
+    ||| Mu
+    TMu   :             Vect k (Name, TDef (S n)) -> TDef n
 
-  ||| The unit type
-  T1    :                                      TDef n
+    ||| Application of a named type to a vector of arguments.
+    TApp  : TNamed k -> Vect k (TDef n)           -> TDef n
 
-  ||| The coproduct type
-  TSum  :         Vect (2 + k) (TDef n)     -> TDef n
+  ||| Named type definition.
+  ||| @n The number of type variables in the type.
+  data TNamed : (n : Nat) -> Type where
+    TName : Name -> TDef n -> TNamed n
```

Comment:
I have the idea that it would be "nice" to avoid the mutual recursion of the name-type and the type-type by letting the name-type be parametric in the type-type, like so:

```idris
||| Named things that are indexed by a `Nat`. (For example `TDef`s.)
||| @thing The type of things to name.
||| @n The 
data Named : (thing : Nat -> Type) -> (n : Nat) -> Type where
  Name : Name -> thing n -> Named thing n

||| Type definition
||| @n The number of type variables in the type
data TDef : (n:Nat) -> Type where
  [...]
  ||| Application of a named type to a vector of arguments.
  TApp  : Named TDef k -> Vect k (TDef n)           -> TDef n
```

It seems to work very well, but except for the "nice feeling" of not having mutually recursive data types, we don't seem to gain very much because none of the mutually recursive functions dealing with these types are easy to turn non-mutually recursive.